File: tests/lua/stream-zip.lua
do
  local function Stream(a) return { __tag = "Stream", a } end
  local function Some(a) return { __tag = "Some", a } end
  local None = { __tag = "None" }
  local Done = { __tag = "Done" }
  local function Yield(a) return { __tag = "Yield", a } end
  local function Skip(a) return { __tag = "Skip", a } end
  local print = print
  local to_string = tostring
  local function zip(tmp)
    local tmp0 = tmp[1]
    local f, start = tmp0._1, tmp0._2
    return function(tmp1)
      local tmp2 = tmp1[1]
      local g = tmp2._1
      return Stream({
        _1 = function(tmp3)
          local tmp4 = tmp3._2
          local sb = tmp4._1
          local x = tmp4._2
          local sa = tmp3._1
          if x.__tag == "None" then
            local tmp5 = f(sa)
            if tmp5.__tag == "Skip" then
              return Skip({ _2 = { _1 = sb, _2 = None }, _1 = tmp5[1] })
            elseif tmp5.__tag == "Yield" then
              local tmp6 = tmp5[1]
              return Skip({ _2 = { _1 = sb, _2 = Some(tmp6._1) }, _1 = tmp6._2 })
            elseif tmp5.__tag == "Done" then
              return Done
            end
          else
            local tmp5 = g(sb)
            local x0 = x[1]
            if tmp5.__tag == "Skip" then
              return Skip({ _1 = sa, _2 = { _1 = tmp5[1], _2 = Some(x0) } })
            elseif tmp5.__tag == "Yield" then
              local tmp6 = tmp5[1]
              return Yield({
                _1 = { _1 = x0, _2 = tmp6._1 },
                _2 = { _1 = sa, _2 = { _1 = tmp6._2, _2 = None } }
              })
            elseif tmp5.__tag == "Done" then
              return Done
            end
          end
        end,
        _2 = { _1 = start, _2 = { _1 = tmp2._2, _2 = None } }
      })
    end
  end
  local tmp = zip(Stream({
    _1 = function(n)
      if n > 100 then return Done end
      return Yield({ _1 = n, _2 = n + 1 })
    end,
    _2 = 1
  }))(Stream({
    _1 = function(n)
      if n > 300 then return Done end
      return Yield({ _1 = n, _2 = n + 1 })
    end,
    _2 = 100
  }))[1]
  local go = tmp._1
  local function go0(ac, st)
    local tmp0 = go(st)
    if tmp0.__tag == "Skip" then
      return go0(ac, tmp0[1])
    elseif tmp0.__tag == "Yield" then
      local tmp1 = tmp0[1]
      local x = tmp1._1
      return go0(x._1 + x._2 + ac, tmp1._2)
    elseif tmp0.__tag == "Done" then
      return ac
    end
  end
  print(to_string(go0(0, tmp._2)))
end

File: tests/lua/stream-zip.ml
external val ( + ) : int -> int -> int = "function(x, y) return x + y end"
external val ( * ) : int -> int -> int = "function(x, y) return x * y end"
external val ( ^ ) : string -> string -> string = "function(x, y) return x .. y end"
external val ( > ) : int -> int -> bool = "function(x, y) return x > y end"
external val ( <= ) : int -> int -> bool = "function(x, y) return x > y end"
external val io_write : string -> unit = "io.write"
external val print : string -> unit = "print"
external val to_string : 'a -> string = "tostring"
external val rem : int -> int -> int = "function(x, y) return x % y end"

type step 's 'a =
  | Skip of 's
  | Yield of 'a * 's
  | Done

type stream 'a =
  | Stream : forall 's. ('s -> step 's 'a) * 's -> stream 'a

type option 'a = Some of 'a | None

let take len (Stream (f, start)) =
  let go (i, st) =
    if i <= len then
      match f st with
      | Done -> Done
      | Skip s -> Skip (i + 1, s)
      | Yield (a, s) -> Yield (a, i + 1, s)
    else
      Done
  Stream (go, 1, start)

let map f (Stream (go, start)) =
  let loop st =
    match go st with
    | Done -> Done
    | Skip s -> Skip s
    | Yield (x, s) -> Yield (f x, s)
  Stream (loop, start)

let filter p (Stream (f, start)) =
  let go st =
    match f st with
    | Done -> Done
    | Skip s -> Skip s
    | Yield (a, s) ->
      if p a then
        Yield (a, s)
      else
        Skip s
  Stream (go, start)

let range (start, limit) =
  let go n =
    if n > limit then
      Done
    else
      Yield (n, n + 1)
  Stream (go, start)

let fold_stream f z (Stream (stream, start)) =
  let rec go ac st =
    match stream st with
    | Yield (a, st) -> go (f a ac) st
    | Skip st -> go ac st
    | Done -> ac
  go z start

let dump_stream e (Stream (f, start)) =
  let rec go st =
    match f st with
    | Skip st -> go st
    | Yield (a, st) ->
      io_write ("'" ^ e a ^ "', ");
      go st
    | Done -> print "]"
  io_write "[";
  go start

let zip (Stream (f, start)) (Stream (g, start')) =
  let go (sa, sb, x) =
    match x with
    | None ->
        match f sa with
        | Done -> Done
        | Skip sa' -> Skip (sa', sb, None)
        | Yield (x, sa') -> Skip (sa', sb, Some x)
    | Some x ->
        match g sb with
        | Done -> Done
        | Skip sb' -> Skip (sa, sb', Some x)
        | Yield (y, sb') -> Yield ((x, y), sa, sb', None)
  Stream (go, start, start', None)

let ( |> ) x f = f x
let uncurry f (x, y) = f x y

let (>>>) f g = fun x -> g (f x)

let sum_squares xs =
  xs |> map (fun x -> x * x) |> fold_stream (+) 0

let () =
  zip (range (1, 100)) (range (100, 300))
        |> map (uncurry (+))
        |> fold_stream (+) 0
        |> to_string
        |> print

